### PR TITLE
#123 merge keys on content_tag

### PR DESCRIPTION
**Pull request by Kr00lIX**
Combining html attributes with the same keys

### CODE REVIEW

**Comment by michalmuskala** [comment]
If we really want to do this, I think it should be done in the `build_attrs` as a single pass over the attributes - we could build a map instead of the list there.

**Comment by josevalim** [comment]
Thank you for the PR!

The issue with this approach is that it is not a generally applicable rule. We can see how data and class are handled in very different ways and each attribute will add its own logic (or require overriding, such as `:id`).

**Comment by Kr00lIX** [comment]
If we still want to do it for the `class`, I can rewrite this code.

I will add this changes to `build_attrs` method and change accumulator to a map

**Comment by Kr00lIX** [comment]
@michalmuskala is it ok now?

File: lib/phoenix_html/tag.ex [hunk 1]
```diff
@@ -99,22 +100,31 @@ defmodule Phoenix.HTML.Tag do
...
-
-  defp build_attrs(_tag, [], acc),
-    do: acc |> Enum.sort |> tag_attrs
+  defp build_attrs(tag, attrs), do: build_attrs(tag, attrs, %{})
+
+  defp build_attrs(_tag, [], acc) do
+    Enum.map(acc, fn
+      {k, v} when is_atom(k) -> {Atom.to_string(k), v}
+      {k, v} when is_bitstring(k) -> {k, v}
+    end)
```
Comment (michalmuskala) [inline lib/phoenix_html/tag.ex]: Why is this conversion needed, it wasn't there before, I believe?

File: lib/phoenix_html/tag.ex [hunk 2]
```diff
@@ -123,12 +133,13 @@ defmodule Phoenix.HTML.Tag do
...
   end
   defp build_attrs(tag, [{k, v}|t], acc) do
-    build_attrs(tag, t, [{dasherize(k), v}|acc])
+    build_attrs(tag, t, Map.put(acc, dasherize(k), v))
   end
 
   defp dasherize(value) when is_atom(value),   do: dasherize(Atom.to_string(value))
   defp dasherize(value) when is_binary(value), do: String.replace(value, "_", "-")
 
+
```
Comment (michalmuskala) [inline lib/phoenix_html/tag.ex]: ✂️

File: lib/phoenix_html/tag.ex (hunk 1 above)
Comment (michalmuskala) [inline lib/phoenix_html/tag.ex]: If anything, we should test for `is_binary`, arbitrary bitstrings would probably crash somewhere down the road

File: lib/phoenix_html/tag.ex [hunk 3]
```diff
@@ -99,22 +100,31 @@ defmodule Phoenix.HTML.Tag do
...
+  defp build_attrs(_tag, [], acc) do
+    Enum.map(acc, fn
+      {k, v} when is_atom(k) -> {Atom.to_string(k), v}
+      {k, v} when is_bitstring(k) -> {k, v}
+    end)
+    |> Enum.sort |> tag_attrs
+  end
+  defp build_attrs(tag, [{k, v}|t], acc) when k in @tag_concat do
+    acc = Map.update(acc, k, v, fn(cv) -> [cv, ' ', v] end)
```
Comment (michalmuskala) [inline lib/phoenix_html/tag.ex]: We can use `[cv, ?\s | v]` as a more memory-efficient way to store this as iodata.

File: lib/phoenix_html/tag.ex (hunk 1 above)
Comment (Kr00lIX) [inline lib/phoenix_html/tag.ex]: Oops, there was my mistake. Thank you

File: lib/phoenix_html/tag.ex (hunk 3 above)
Comment (Kr00lIX) [inline lib/phoenix_html/tag.ex]: Thanks, I will know that for the future

**Comment by Kr00lIX** [comment]
Thanks for the review, I fixed it

**Comment by michalmuskala** [comment]
Now it's up to somebody that could actually merge the PR 😆

**Comment by Kr00lIX** [comment]
Thanks for the review @josevalim ! Pushed an update for all the things you mentioned.

**Comment by josevalim** [comment]
This seems to be a backwards incompatible solution if some attributes can be repeated. Is there any attribute that can behave as such?

**Comment by kelvinst** [comment]
Well, never heard about the possibility to set the same attribute multiple times in HTML. But also, I could not find anything on HTML specification that defines what happens then. :/

Testing a little bit here, on the cases like:

```html
<a href="/portfolios" href="/portfolios2" id="this-el">Portfolios</a>
```

When you get the value of href with:

```js
document.getElementById("this-el").href
// => "http://localhost:4000/portfolios"
```

Looks like the browser uses the first informed value of the property, which is actually the inverse of what you're doing when you merge the new attribute on the current map.

I particularly like this PR idea because it makes sense to our heads, but considering it breaks the default browser behavior, I think this ideas is not as good as it looks like.

**Comment by josevalim** [comment]
Sorry for the delay on this. Since I am still not convinced we want to generally perform this change, I will go ahead and close this. Thank you!

**Pull request closed without merging**